Xamarin.Essentials/TextToSpeech/TextToSpeech.mac.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using AppKit;
+
+namespace Xamarin.Essentials
+{
+    public static partial class TextToSpeech
+    {
+        static readonly Lazy<NSSpeechSynthesizer> speechSynthesizer = new Lazy<NSSpeechSynthesizer>(() =>
+            new NSSpeechSynthesizer { Delegate = new SpeechSynthesizerDelegate() });
+
+        internal static Task<IEnumerable<Locale>> PlatformGetLocalesAsync() =>
+            Task.FromResult(NSSpeechSynthesizer.AvailableVoices
+                .Select(v => new Locale(v, null, null, null)));
+
+        internal static async Task PlatformSpeakAsync(string text, SpeechOptions options, CancellationToken cancelToken = default)
+        {
+            var ss = speechSynthesizer.Value;
+            var ssd = (SpeechSynthesizerDelegate)ss.Delegate;
+
+            var tcs = new TaskCompletionSource<bool>();
+            try
+            {
+                // Ensures linker doesn't remove.
+                if (DateTime.UtcNow.Ticks < 0)
+                {
+                    new NSSpeechSynthesizer();
+                    new SpeechSynthesizerDelegate();
+                }
+
+                if (options != null)
+                {
+                    if (options.Volume.HasValue)
+                        ss.Volume = NormalizeVolume(options.Volume);
+
+                    if (options.Locale != null)
+                        ss.Voice = options.Locale.Language;
+                }
+
+                ssd.FinishedSpeaking += OnFinishedSpeaking;
+                ssd.EncounteredError += OnEncounteredError;
+
+                ss.StartSpeakingString(text);
+
+                using (cancelToken.Register(TryCancel))
+                {
+                    await tcs.Task;
+                }
+            }
+            finally
+            {
+                ssd.FinishedSpeaking -= OnFinishedSpeaking;
+                ssd.EncounteredError -= OnEncounteredError;
+            }
+
+            void TryCancel()
+            {
+                ss.StopSpeaking(NSSpeechBoundary.hWord);
+                tcs.TrySetResult(true);
+            }
+
+            void OnFinishedSpeaking(bool completed)
+            {
+                tcs.TrySetResult(completed);
+            }
+
+            void OnEncounteredError(string errorMessage)
+            {
+                // TODO: a real exception type here
```

Comment:
Hack?

Reply:
No, we are using `Exception` in places, not the best. @Redth @jamesmontemagno this is sort of where we need to actually look in the rest of the library.